### PR TITLE
Feature/graph DFS Finder refactor smart server provider

### PR DESCRIPTION
**Pull request by hkernbach**
### Scope & Purpose

First implementation of OneSidedTraverser including DFSPathFinder and Tests.
The source code is also prepared to run a BFSPathFinder (use of different FiFo queue instead of LiFo queue), but this is not finalized and not yet tested.

- [x] :pizza: Refactoring

#### Backports:

- [x] No backports required

### Testing & Verification

- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**

### CODE REVIEW

File: arangod/Graph/Enumerators/OneSidedEnumerator.cpp [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+auto OneSidedEnumerator<QueueType, PathStoreType, ProviderType, PathValidator>::destroyEngines()
+    -> void {
+  _provider.destroyEngines();
+}
+
+template <class QueueType, class PathStoreType, class ProviderType, class PathValidator>
+void OneSidedEnumerator<QueueType, PathStoreType, ProviderType, PathValidator>::clear() {
+  _interior.reset();
+  _queue.clear();
+  _currentDepth = 0;
```
Comment (hkernbach) [inline arangod/Graph/Enumerators/OneSidedEnumerator.cpp:74]:
```suggestion
```

File: arangod/Graph/Enumerators/OneSidedEnumerator.cpp [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+template <class QueueType, class PathStoreType, class ProviderType, class PathValidator>
+OneSidedEnumerator<QueueType, PathStoreType, ProviderType, PathValidator>::OneSidedEnumerator(
+    ProviderType&& forwardProvider, OneSidedEnumeratorOptions&& options,
+    arangodb::ResourceMonitor& resourceMonitor)
+    : _options(std::move(options)),
+      _queue(resourceMonitor),
+      _provider(std::move(forwardProvider)),
+      _validator(_interior),
+      _interior(resourceMonitor),
+      _resultPath{_provider, _provider} {}
```
Comment (mchacki) [inline arangod/Graph/Enumerators/OneSidedEnumerator.cpp:58]: We only need 1 Provider, I am confused.

File: arangod/Graph/Providers/SingleServerProvider.h [hunk 3]
```diff
@@ -65,9 +65,17 @@ struct SingleServerProvider {
    public:
     class Vertex {
      public:
-      explicit Vertex(VertexType v) : _vertex(v) {}
+      explicit Vertex(VertexType v) : _vertex(v) {
```
Comment (mchacki) [inline arangod/Graph/Providers/SingleServerProvider.h:68]: The Vertex does not need the Depth.
Only the Step requires it.

File: arangod/Graph/Queues/LifoQueue.h [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    Step first = std::move(_queue.front());
+    _resourceMonitor.decreaseMemoryUsage(sizeof(Step));
+    LOG_TOPIC("9cd65", TRACE, Logger::GRAPHS) << "<LifoQueue> Pop: " << first.toString();
+    _queue.pop_front();
+    return first;
+  }
+
+ private:
+  /// @brief queue datastore
+  std::deque<Step> _queue;
```
Comment (mchacki) [inline arangod/Graph/Queues/LifoQueue.h:95]: I think std::stack should have better performance here.

File: arangod/Graph/algorithm-aliases.h [hunk 5]
```diff
@@ -23,6 +23,12 @@
 
 #pragma once
 
+#include <Graph/Enumerators/OneSidedEnumerator.h>
```
Comment (mchacki) [inline arangod/Graph/algorithm-aliases.h:26]:
```suggestion
#include "Graph/Enumerators/OneSidedEnumerator.h"
```

for all includes in this file

**Review by mchacki — CHANGES_REQUESTED** [review]
I have added some comments on Production.
I have not yet reviewed the Test Files.

File: tests/Graph/MockGraphProvider.h [hunk 6]
```diff
@@ -67,9 +67,11 @@ class MockGraphProvider {
    public:
     class Vertex {
      public:
-      explicit Vertex(VertexType v) : _vertex(v){};
+      explicit Vertex(VertexType v) : _vertex(v) { _depth = 0; };
```
Comment (mchacki) [inline tests/Graph/MockGraphProvider.h:70]: Also here I think the Vertex does not have to know it's depth, the Step is enough

**Review by mchacki — CHANGES_REQUESTED** [review]
Added comments within tests.

In general good.
One Test is wrong.
I am Missing LIFO queue tests.

File: arangod/Graph/Enumerators/OneSidedEnumerator.cpp [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    auto&& preparedEnds = futureEnds.get();
+
+    TRI_ASSERT(preparedEnds.size() != 0);
+    TRI_ASSERT(_queue.hasProcessableElement());
+  }
+
+  auto step = _queue.pop();
+  auto posPrevious = _interior.append(step);
+
+  if (!step.isFirst()) {
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply arangod/Graph/Enumerators/OneSidedEnumerator.cpp]: Thanks for spotting. Changed!

File: arangod/Graph/Enumerators/OneSidedEnumerator.cpp (hunk 2 above)
Comment (hkernbach) [reply arangod/Graph/Enumerators/OneSidedEnumerator.cpp]: Implemented new ResultType (which only needs one provider)

File: arangod/Graph/Providers/SingleServerProvider.h (hunk 3 above)
Comment (hkernbach) [reply arangod/Graph/Providers/SingleServerProvider.h:68]: Moved to BaseStep and one level above ( -> Step)

File: arangod/Graph/algorithm-aliases.h (hunk 5 above)
Comment (hkernbach) [reply arangod/Graph/algorithm-aliases.h]: Ty, done.

File: tests/Graph/DFSFinderTest.cpp [hunk 8]
```diff
@@ -0,0 +1,579 @@
...
+
+  // Source
+  auto source = vId(1);
+
+  finder.reset(toHashedStringRef(source));
+
+  EXPECT_FALSE(finder.isDone());
+  {
+    result.clear();
+    auto hasPath = finder.getNextPath(result);
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/Graph/DFSFinderTest.cpp:285]: Should be fixed now. Will keep this here open until you've checked.

File: tests/Graph/MockGraphProvider.h (hunk 6 above)
Comment (hkernbach) [reply tests/Graph/MockGraphProvider.h]: Changed and moved to parent class:  BaseStep.

File: tests/Graph/DFSFinderTest.cpp [hunk 9]
```diff
@@ -0,0 +1,579 @@
...
+    EXPECT_TRUE(hasPath);
+
+    pathStructureValid(result.slice(), 1);
+    pathEquals(result.slice(), {1, 2});
+
+    EXPECT_TRUE(finder.isDone()); // TODO CHECK - changed to EXPECT_TRUE
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/Graph/DFSFinderTest.cpp:324]: +1, should keep this in mind when it comes to filter testing :)

File: arangod/Graph/Queues/LifoQueue.h (hunk 4 above)
Comment (hkernbach) [reply arangod/Graph/Queues/LifoQueue.h:95]: As discussed, currently we'll continue to use deque.

File: tests/Graph/DFSFinderTest.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/Graph/DFSFinderTest.cpp]: > I am Missing LIFO queue tests.

Added.

File: arangod/Graph/PathManagement/SingleProviderPathResult.cpp [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Michael Hackstein
+/// @author Heiko Kernbach
+////////////////////////////////////////////////////////////////////////////////
+
+#include "./SingleProviderPathResult.h"
```
Comment (mchacki) [inline arangod/Graph/PathManagement/SingleProviderPathResult.cpp:25]:
```suggestion
#include "SingleProviderPathResult.h"
```

File: arangod/Graph/PathManagement/SingleProviderPathResult.cpp [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+template class ::arangodb::graph::SingleProviderPathResult<::arangodb::graph::ProviderTracer<::arangodb::graph::SingleServerProvider>,
+                                                   ::arangodb::graph::SingleServerProvider::Step>;
+
+/* ClusterProvider Section */
+
+template class ::arangodb::graph::SingleProviderPathResult<::arangodb::graph::ClusterProvider, ::arangodb::graph::ClusterProvider::Step>;
+
+template class ::arangodb::graph::SingleProviderPathResult<::arangodb::graph::ProviderTracer<::arangodb::graph::ClusterProvider>,
+                                                   ::arangodb::graph::ClusterProvider::Step>;
```
Comment (mchacki) [inline arangod/Graph/PathManagement/SingleProviderPathResult.cpp:110]:
```suggestion
                                                   ::arangodb::graph::ClusterProvider::Step>;

```

File: arangod/Graph/PathManagement/SingleProviderPathResult.h [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+ private:
+  std::vector<typename Step::Vertex> _vertices;
+  std::vector<typename Step::Edge> _edges;
+  
+  // Provider for the path
+  ProviderType& _provider;
+};
+}  // namespace graph
+}  // namespace arangodb
```
Comment (mchacki) [inline arangod/Graph/PathManagement/SingleProviderPathResult.h:63]:
```suggestion
}  // namespace arangodb

```

**Review by mchacki — APPROVED** [review]
LGTM.
Just some style modifications

File: arangod/Graph/PathManagement/SingleProviderPathResult.cpp (hunk 11 above)
Comment (hkernbach) [reply arangod/Graph/PathManagement/SingleProviderPathResult.cpp]: -> clang format applied now

File: arangod/Graph/Enumerators/OneSidedEnumerator.cpp [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+    ValidationResult res = _validator.validatePath(step);
+    if ((step.getDepth() >= _options.getMinDepth()) && !res.isFiltered() && !res.isPruned()) {
+      // Include it in results.
+      _results.push_back(std::make_pair(_interior.get(posPrevious), step));
+    }
+  }
+
+  if (step.getDepth() < _options.getMaxDepth()) {
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply arangod/Graph/Enumerators/OneSidedEnumerator.cpp]: Will do this in graph refactor. Card created.

**Comment by mchacki** [comment]
Red test is unrelated.

**Pull request merged**